File: Android/Slime/src/gamers/associate/Slime/CameraManager.java
package gamers.associate.Slime;

import org.cocos2d.layers.CCLayer;
import org.cocos2d.nodes.CCDirector;
import org.cocos2d.types.CGPoint;
import org.cocos2d.types.CGRect;

public class CameraManager {
	
	private float minScale;
	private float maxScale;
	private float zoomSpeed;
	private float screenW2 = CCDirector.sharedDirector().winSize().width * CCDirector.sharedDirector().winSize().width;
	private float screenH2 = CCDirector.sharedDirector().winSize().height * CCDirector.sharedDirector().winSize().height;
	
	private CCLayer gameLayer;
	private float levelWidth;
	private float levelHeight;	
	private boolean isCameraOnContinuousMove;
	private CGPoint moveCameraBy;
	private CGRect cameraView;
	private CGPoint virtualCameraPos;
	private GameItem followed;	
	private CGPoint levelOrigin;
	
	public CameraManager(CCLayer gameLayer) {
		this.gameLayer = gameLayer;
				
		this.moveCameraBy = new CGPoint();
		this.virtualCameraPos = CGPoint.getZero();
	}
	
	protected void tick(float delta) {
		if (this.isCameraOnContinuousMove) {
			this.moveCameraBy(this.moveCameraBy);
		}
		
		if (this.followed != null) {
			this.centerCameraOn(this.followed.getPosition());
		}
	}
	
	public void follow(GameItem item) {
		this.followed = item;
	}
		
	public void normalizePosition() {
		float scale = this.gameLayer.getScale();		
		boolean isNormalized = false;
		CGPoint position = this.gameLayer.getPosition();
		
		float maxLeft = CGRect.minX(this.cameraView);
		float left = position.x;
		if (left > maxLeft) {
			position.x += maxLeft - left;
			isNormalized = true;
		}
		
		float minRight = CGRect.maxX(this.cameraView);
		float right = position.x + this.levelWidth * scale;
		if (right < minRight) {
			position.x += minRight - right;
			isNormalized = true;
		}
		
		float maxBottom = CGRect.minY(this.cameraView);
		float bottom = position.y;
		if (bottom > maxBottom) {
			position.y += maxBottom - bottom;
			isNormalized = true;
		}
		
		float minTop = CGRect.maxY(this.cameraView);
		float top = position.y + this.levelHeight * scale;
		if (top < minTop) {
			position.y += minTop - top;
			isNormalized = true;
		}						
		
		this.gameLayer.setPosition(position);
		this.virtualCameraPos.set(- position.x, - position.y);
		if (isNormalized) {
			this.isCameraOnContinuousMove = false;
		}
	}
	
	public void centerCameraOn(CGPoint center) {		
		this.keepPointAt(center, CGPoint.make(this.cameraView.size.width / 2, this.cameraView.size.height / 2));
	}
	
	public void keepPointAt(CGPoint gamePoint, CGPoint screenPin) {
		float scale = this.gameLayer.getScale();
		CGPoint position = CGPoint.make(
				- (gamePoint.x * scale - screenPin.x),
				- (gamePoint.y * scale - screenPin.y));
		this.setLayerPosition(position);
	}
	
	private void setLayerPosition(CGPoint position) {				
		this.gameLayer.setPosition(position);
		this.normalizePosition();
	}
	
	public void setCameraPosition(CGPoint origin) {
		float scale = this.gameLayer.getScale();
		CGPoint position = CGPoint.make(
				- origin.x * scale,
				- origin.y * scale);
		this.setLayerPosition(position);
	}
	
	public void moveCameraBy(CGPoint move) {
		float scale = this.gameLayer.getScale();
		CGPoint position = CGPoint.make(this.gameLayer.getPosition().x, this.gameLayer.getPosition().y);
		position.x += - move.x * scale;
		position.y += - move.y * scale;
		
		this.setLayerPosition(position);
	}
	
	public void continuousMoveCameraBy(CGPoint delta) {
		// TODO: must be fix based on time (/sec)
		float maxMovePos = 10f;
		float maxMoveNeg = -10f;
		float minMove = 1.0f;
		
		if (Math.abs(delta.x) > minMove) {
			if (delta.x > maxMovePos) delta.x = maxMovePos;
			if (delta.x < maxMoveNeg) delta.x = maxMoveNeg;
		}
		else {
			delta.x = 0f;		
		}
		
		if (Math.abs(delta.y) > minMove) {
			if (delta.y > maxMovePos) delta.y = maxMovePos;
			if (delta.y < maxMoveNeg) delta.y = maxMoveNeg;		
		}
		else {
			delta.y = 0f;
		}
		
		this.moveCameraBy.x = delta.x;
		this.moveCameraBy.y = delta.y;
		this.isCameraOnContinuousMove = true;
	}
	
	public void stopContinousMoving() {
		this.isCameraOnContinuousMove = false;
	}			
	
	public void zoomCameraByScreenRatio(float zoomDelta) {								
		double max = Math.sqrt(screenW2 + screenH2);
		float zoom = (float) (zoomDelta * maxScale / max) * this.zoomSpeed;
		this.zoomCameraBy(zoom);
	}
	
	private CGPoint zoomAnchor;
	private CGPoint zoomScreenPin;
	
	public void setZoomPoint(CGPoint zoomPoint) {
		this.zoomScreenPin = CGPoint.make(zoomPoint.x, zoomPoint.y);
		float scale = this.gameLayer.getScale();
		CGPoint zoomScaled = CGPoint.zero();
		zoomScaled.x =  (zoomPoint.x - this.gameLayer.getPosition().x) / scale; 
		zoomScaled.y = (zoomPoint.y - this.gameLayer.getPosition().y) / scale;
		
		this.zoomAnchor = zoomScaled;
	}
	
	public void zoomCameraBy(float zoomDelta) {		
		float scale = this.gameLayer.getScale() + zoomDelta;
		this.zoomCameraTo(scale);
	}
	
	public void zoomCameraTo(float zoomValue) {
		if (zoomValue <= minScale) {
			zoomValue = minScale;
		}
		
		if (zoomValue >= maxScale) {
			zoomValue = maxScale;
		}
		
		this.gameLayer.setScale(zoomValue);
		this.keepPointAt(this.zoomAnchor, this.zoomScreenPin);
		this.normalizePosition();
	}
	
	public void setCameraView() {		
		CGPoint origin = this.levelOrigin;		
		// Default view is windows size
		this.cameraView = CGRect.make(origin, CCDirector.sharedDirector().winSize());
		
		// By default camera can not be unzoom more than the most little size
		this.zoomAnchor = CGPoint.getZero();
		this.zoomScreenPin = CGPoint.getZero();
		float minScaleW = 1 / (this.levelWidth / CGRect.width(this.cameraView));
		float minScaleH = 1 / (this.levelHeight / CGRect.height(this.cameraView));
		this.minScale = Math.max(minScaleW, minScaleH);
		this.maxScale = 2.0f;
		this.zoomSpeed = 3.0f;
		// To take into account new limits
		this.zoomCameraBy(0);
		this.normalizePosition();
	}
	
	public void attachLevel(float levelWidth, float levelHeight, CGPoint levelOrigin) {
		this.levelWidth = levelWidth;
		this.levelHeight = levelHeight;				
		this.levelOrigin = levelOrigin;
	}
}

File: Iphone/Slime/Classes/CameraManager.m
#import "CameraManager.h"


@implementation CameraManager


- (id) init:(CCLayer *)my_gameLayer levelWidth:(float)my_levelWidth levelHeight:(float)my_levelHeight levelOrigin:(CGPoint )my_levelOrigin {
	self = [super init];
	if (self != nil) {
    screenW2 = [[CCDirector sharedDirector] winSize].width * [[CCDirector sharedDirector] winSize].width;
    screenH2 = [[CCDirector sharedDirector] winSize].height * [[CCDirector sharedDirector] winSize].height;
    gameLayer = my_gameLayer;
    levelWidth = my_levelWidth;
    levelHeight = my_levelHeight;
    levelOrigin = my_levelOrigin;
    //moveCameraBy = [[[CGPoint alloc] init] autorelease];
    virtualCameraPos = CGPointZero;
    [self setCameraView];
  }
  return self;
}

- (void) tick:(float)delta {
  if (isCameraOnContinuousMove) {
    [self moveCameraBy:moveCameraBy];
  }
  if (followed != nil) {
    [self centerCameraOn:[followed position]];
  }
}

- (void) follow:(GameItem *)item {
  followed = item;
}

- (void) normalizePosition {
  float scale = [gameLayer scale];
  BOOL isNormalized = NO;
  CGPoint  position = [gameLayer position];
  float maxLeft =CGRectGetMinX(cameraView);
  float left = position.x;
  if (left > maxLeft) {
    position.x += maxLeft - left;
    isNormalized = YES;
  }
  float minRight = CGRectGetMaxX(cameraView);
  float right = position.x + levelWidth * scale;
  if (right < minRight) {
    position.x += minRight - right;
    isNormalized = YES;
  }
  float maxBottom = CGRectGetMinY(cameraView);
  float bottom = position.y;
  if (bottom > maxBottom) {
    position.y += maxBottom - bottom;
    isNormalized = YES;
  }
  float minTop = CGRectGetMaxY(cameraView);
  float top = position.y + levelHeight * scale;
  if (top < minTop) {
    position.y += minTop - top;
    isNormalized = YES;
  }
  [gameLayer setPosition:position];
  virtualCameraPos = ccp(-position.x,-position.y);
  if (isNormalized) {
    isCameraOnContinuousMove = NO;
  }
}

- (void) centerCameraOn:(CGPoint )center {
  [self keepPointAt:center screenPin:ccp(cameraView.size.width / 2 ,cameraView.size.height / 2)];
}

- (void) keepPointAt:(CGPoint )gamePoint screenPin:(CGPoint)screenPin {
  float scale = [gameLayer scale];
  CGPoint  position = ccp(gamePoint.x * scale - screenPin.x, gamePoint.y * scale - screenPin.y);
  [self setLayerPosition:position];
}

- (void) setLayerPosition:(CGPoint )position {
  [gameLayer setPosition:position];
  [self normalizePosition];
}

- (void) setCameraPosition:(CGPoint )origin {
  float scale = [gameLayer scale];
  CGPoint  position = ccp(-origin.x * scale, -origin.y * scale);
  [self setLayerPosition:position];
}

- (void) moveCameraBy:(CGPoint )move {
  float scale = [gameLayer scale];
  CGPoint  position = ccp([gameLayer position].x , [gameLayer position].y);
  position.x += -move.x * scale;
  position.y += -move.y * scale;
  [self setLayerPosition:position];
}

- (void) continuousMoveCameraBy:(CGPoint )delta {
  float maxMovePos = 10.0f;
  float maxMoveNeg = -10.0f;
  float minMove = 1.0f;
  if (abs(delta.x) > minMove) {
    if (delta.x > maxMovePos)
      delta.x = maxMovePos;
    if (delta.x < maxMoveNeg)
      delta.x = maxMoveNeg;
  }
   else {
    delta.x = 0.0f;
  }
  if (abs(delta.y) > minMove) {
    if (delta.y > maxMovePos)
      delta.y = maxMovePos;
    if (delta.y < maxMoveNeg)
      delta.y = maxMoveNeg;
  }
   else {
    delta.y = 0.0f;
  }
  moveCameraBy.x = delta.x;
  moveCameraBy.y = delta.y;
  isCameraOnContinuousMove = YES;
}

- (void) stopContinousMoving {
  isCameraOnContinuousMove = NO;
}

- (void) zoomCameraByScreenRatio:(float)zoomDelta {
  double max = sqrt(screenW2 + screenH2);
  float zoom = (float)(zoomDelta * maxScale / max) * zoomSpeed;
  [self zoomCameraBy:zoom];
}

- (void) setZoomPoint:(CGPoint )zoomPoint {
  zoomScreenPin = ccp(zoomPoint.x, zoomPoint.y);
  float scale = [gameLayer scale];
  CGPoint zoomScaled = CGPointZero;
  zoomScaled.x = (zoomPoint.x - [gameLayer position].x) / scale;
  zoomScaled.y = (zoomPoint.y - [gameLayer position].y) / scale;
  zoomAnchor = zoomScaled;
}

- (void) zoomCameraBy:(float)zoomDelta {
  float scale = [gameLayer scale] + zoomDelta;
  [self zoomCameraTo:scale];
}

- (void) zoomCameraTo:(float)zoomValue {
  if (zoomValue <= minScale) {
    zoomValue = minScale;
  }
  if (zoomValue >= maxScale) {
    zoomValue = maxScale;
  }
  [gameLayer setScale:zoomValue];
  [self keepPointAt:zoomAnchor screenPin:zoomScreenPin];
  [self normalizePosition];
}

- (void) setCameraView {
  CGPoint  origin = levelOrigin;
  cameraView = CGRectMake(origin.x, origin.y ,[[CCDirector sharedDirector] winSize].width , [[CCDirector sharedDirector] winSize].height);
  zoomAnchor = CGPointZero;
  zoomScreenPin = CGPointZero;
  float minScaleW = 1 / (levelWidth / CGRectGetWidth(cameraView));
  float minScaleH = 1 / (levelHeight / CGRectGetHeight(cameraView));
  minScale = MAX(minScaleW, minScaleH);
  maxScale = 2.0f;
  zoomSpeed = 3.0f;
  [self zoomCameraBy:0];
  [self normalizePosition];
}

- (void) dealloc {
  [gameLayer release];
 
  [followed release];
    [super dealloc];
}

@end
